libvast/src/format/syslog.cpp

```diff
@@ -12,9 +12,148 @@
+  
+  // Syslog structured_data type.
+  // There can be multiple key value pairs for one structured data element
+  auto structured_data_fields = std::vector<record_field>{
+    {"sd_id", string_type{}},
+    {"sd_elements", string_type{}},
+  }; 
+  types_.structured_data_type = record_type{std::move(structured_data_fields)}.name(
+    "syslog.strucured_data");
+
+  if (in != nullptr)
+    reset(std::move(in));
+}
+
+void reader::reset(std::unique_ptr<std::istream> in) {
+  VAST_ASSERT(in != nullptr);
+  input_ = std::move(in);
+  lines_ = std::make_unique<detail::line_range>(*input_);
+}
+
+const char* reader::name() const {
+  return "syslog-reader";
+}
+
+
+caf::error reader::read_impl(size_t max_events, size_t max_slice_size,
+                             consumer& f) {
+  // Sanity checks.
+  VAST_ASSERT(max_events > 0);
+  VAST_ASSERT(max_slice_size > 0);
+
+  if (builder_ == nullptr) {
+      if (!caf::holds_alternative<record_type>(types_.header_type))
+        return make_error(ec::parse_error, "illegal syslog message type");
+      if (!reset_builder(caf::get<record_type>(types_.header_type)))
+        return make_error(ec::parse_error,
+                          "unable to create builder for syslog message type");
+  }
+  VAST_ASSERT(builder_ != nullptr);
+  // EOF check.
+  if (lines_->done())
+    return make_error(ec::end_of_input, "input exhausted");
+  
+  // Local buffer for parsing records.
+  std::vector<data> xs;
+  // Counts successfully parsed records.
+  size_t produced = 0;
+  // Loop until reaching EOF or the configured limit of records.
+  while (produced < max_events) {
+    lines_->next();
+    if (lines_->done())
+      return finish(f, make_error(ec::end_of_input, "input exhausted"));
+    // Parse curent line.
+    auto& line = lines_->get();
+    syslog::syslog_msg sys_msg;
+    if (!parser_(line, sys_msg))
+      return make_error(ec::format_error,
+                                  "failed to parse Syslog message");
+    
+    auto severity = sys_msg.header.priority % 8; // RFC 5424 The Priority value is calculated by first multiplying the Facility number by 8 and then adding the numerical value of the Severity.
+    auto facility = sys_msg.header.priority - severity;
+    if (!(builder_->add(facility) && builder_->add(severity)
+          && builder_->add(sys_msg.header.ts) && builder_->add(sys_msg.header.hostname)
+          && builder_->add(sys_msg.header.app_name)
+          && builder_->add(sys_msg.header.process_id)
+          && builder_->add(sys_msg.header.msg_id)
+          && builder_->add(sys_msg.header.version)
+          && builder_->add(sys_msg.msg.msg_content)
+    )) {
+      return make_error(ec::parse_error, "unable to fill row");
+    }
+    // The layout of the builder needs to changed, would it be better to have to builders? One for the header and message and one for the structured data, or even three (separate header from msg part again)
```

Comment:
One builder for every layout. That means you'll only need a single builder, that you can create at the start of the `read_impl` call.